Fixing sentences and sources (#373)
OWASP/wstg: Testing_for_Vertical_Bypassing_Authorization_Schema_WSTG-AUTHZ-00X.md

## Summary

It is common in modern enterprises to define system roles to manage users and authorization to system resources. The purpose of the system role is to distinguish between user levels and to bind the assigned privileges to each role. An Authorization bypass enables an attacker to gain access to resources that are usually protected from an application or user. A Vertical authorization bypass is specific to the case that an attacker obtains a role higher than his own. This kind of test focuses on verifying how the Vertical authorization schema has been implemented for each role in getting access rights to data and resources of other users with different roles or privileges. Vertical authorization bypass may occur when an attacker gets access to more resources or data than they usually are allowed. The application should prevent such elevation or changes. For every function, page, specific role, and request that the application executes during the post-authentication phase, it is necessary to verify:

-   Is it possible to access resources that should be accessible only to higher role user?

> -   Is it possible to operate functions on resources that should be operative only by a user that holds a higher or specific role identity?

Reviewer: Please revert these bullet spacing changes

> ## How to test

Reviewer: ```suggestion
## How to Test

```

If you want to test manually, the process of testing for bypass authorization scheme could look like this, for each role:

> 1.  For each role, register/generate a user.

Reviewer: Please remove the extra leading spaces added here

## Summary

> It is common in modern enterprises to define system roles to manage users and authorization to system resources. The purpose of the system role is to distinguish between user levels and to bind the assigned privileges to each role. An Authorization bypass enables an attacker to gain access to resources that are usually protected from an application or user. A Vertical authorization bypass is specific to the case that an attacker obtains a role higher than his own. This kind of test focuses on verifying how the Vertical authorization schema has been implemented for each role in getting access rights to data and resources of other users with different roles or privileges. Vertical authorization bypass may occur when an attacker gets access to more resources or data than they usually are allowed. The application should prevent such elevation or changes. For every function, page, specific role, and request that the application executes during the post-authentication phase, it is necessary to verify:

Reviewer 2: Replace `his` with `their` -> `a role higher than their own`

- Is it possible to access resources that should be accessible only to higher role user?
- Is it possible to operate functions on resources that should be operative only by a user that holds a higher or specific role identity?

## How to Test

> If you want to test manually, the process of testing for bypass authorization scheme could look like this, for each role:

Reviewer 2: remove `, for each role` in the end of the line

1. For each role, register/generate a user.
2. Generate and keep the session tokens by authenticating the application (one session token for each user).
3. For every request, change the session token from the original token to another role session token and diagnose the responses for each token.
4. An application will be considered vulnerable if the responses are the same, contain same private data or indicate successful operation on other users resource or data.

> Under the section tools there are mulitple addons for pentesting frameworks OWASP Zed Attack Proxy (ZAP) and Burp, which enable testing the authorization bypass automatically and on a larger scale. Though a review of the process that is used, is recommendable.

Reviewer 2: ```suggestion
In the [Tools](#tools) section, there are multiple add-ons for proxies such as ZAP and Burp that enables them to automatically conduct vertical authorization bypasses on a larger scale, coupling them with manual configuration and report reviewing.
```

| ROLE | PERMISSION | ADDITIONAL PERMISSION |
|------|------------|-----------------------|
| Administrator | Full Control     | Delete |
| Manager       | Modify, Add, Read | Add    |
| Staff         | Read, Modify     | Modify |
| Customer      | Read Only        |        |

The application will be considered vulnerable if:

> 1.  Customer could operate administrator, manager or staff functions;

Reviewer: Please remove the spacing added to these ordered list items

### Example 2

> Suppose that the administrator menu is part of the administrator account. The application will be considered vulnerable if any other role rather than administrator could access the administrator menu. Sometimes, developer perform authorization validation at the GUI level only, and leave the functions without authorization validation, thus resulting possibly in a vulnerability.

Reviewer: ```suggestion
Suppose that the administrator menu is part of the administrator account. The application will be considered vulnerable if any other role rather than administrator could access the administrator menu. Sometimes, developer perform authorization validation at the GUI level only, and leave the functions without authorization validation, thus potentially resulting in a vulnerability.

```

## Tools

- ZAP extension: Access Control Testing. <https://www.zaproxy.org/docs/desktop/addons/access-control-testing/>

> 

Reviewer: ```suggestion
```

- Burp extension: AuthMatrix. <https://github.com/SecurityInnovation/AuthMatrix/>

> - Burp extension: Authorize. <https://github.com/Quitten/Autorize>

Reviewer: ```suggestion
- [ZAP extension: Access Control Testing](https://www.zaproxy.org/docs/desktop/addons/access-control-testing/)
- [Burp extension: AuthMatrix](https://github.com/SecurityInnovation/AuthMatrix/)
- [Burp extension: Authorize](https://github.com/Quitten/Autorize)
```

> -   Burp extension: Authorize. <https://github.com/Quitten/Autorize>

Reviewer: Please remove the blank lines and extra leading spaces
